File: algorithms/julia/kmp.jl
function build_lps(pat::String)
    m   = length(pat)
    lps = zeros(Int, m)
    len, i = 0, 2
    while i <= m
        if pat[i] == pat[len+1]
            len += 1
            lps[i] = len
            i += 1
        elseif len > 0
            len = lps[len]
        else
            lps[i] = 0
            i += 1
        end
    end
    return lps
end

function kmp(text::String, pat::String)
    n, m = length(text), length(pat)
    m == 0 && return Int[]
    lps     = build_lps(pat)
    matches = Int[]
    i, j    = 1, 1
    while i <= n
        if text[i] == pat[j]
            i += 1; j += 1
        end
        if j > m
            push!(matches, i - j)
            j = lps[j-1] + 1
        elseif i <= n && text[i] != pat[j]
            j > 1 ? (j = lps[j-1] + 1) : (i += 1)
        end
    end
    return matches
end

function main()
    lines = readlines(stdin)
    text  = strip(lines[1])
    pat   = strip(lines[2])

    # JIT warmup
    kmp("abab", "ab")

    start   = time_ns()
    matches = kmp(text, pat)
    ms      = (time_ns() - start) / 1e6

    println(length(matches))
    println(isempty(matches) ? "" : join(matches .- 1, " "))
    println(stderr, ms)
end

main()

File: algorithms/go/kmp.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

func buildLPS(pat string) []int {
	m   := len(pat)
	lps := make([]int, m)
	len_, i := 0, 1
	for i < m {
		if pat[i] == pat[len_] {
			len_++
			lps[i] = len_
			i++
		} else if len_ > 0 {
			len_ = lps[len_-1]
		} else {
			lps[i] = 0
			i++
		}
	}
	return lps
}

func kmp(text, pat string) []int {
	n, m := len(text), len(pat)
	if m == 0 { return nil }
	lps     := buildLPS(pat)
	matches := []int{}
	i, j   := 0, 0
	for i < n {
		if text[i] == pat[j] { i++; j++ }
		if j == m {
			matches = append(matches, i-j)
			j = lps[j-1]
		} else if i < n && text[i] != pat[j] {
			if j > 0 { j = lps[j-1] } else { i++ }
		}
	}
	return matches
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	text, _ := reader.ReadString('\n')
	pat,  _  := reader.ReadString('\n')
	text = strings.TrimSpace(text)
	pat  = strings.TrimSpace(pat)

	start   := time.Now()
	matches := kmp(text, pat)
	ms      := float64(time.Since(start).Nanoseconds()) / 1e6

	fmt.Fprintln(writer, len(matches))
	if len(matches) == 0 {
		fmt.Fprintln(writer)
	} else {
		sb := make([]string, len(matches))
		for i, v := range matches { sb[i] = fmt.Sprintf("%d", v) }
		fmt.Fprintln(writer, strings.Join(sb, " "))
	}
	fmt.Fprintf(os.Stderr, "%.3f\n", ms)
}